example/neural_collaborative_filtering/core/evaluate.py

```diff
@@ -0,0 +1,108 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# 
+import math
+import heapq
+import random
+import numpy as np
+import logging
+import mxnet as mx
+
+def get_movielens_iter(filename, batch_size, ctx, logger):
+    """Not particularly fast code to parse the text file and load into NDArrays.
+    return two data iters, one for train, the other for validation.
+    """
+    logger.info("Preparing data iterators for " + filename + " ... ")
+    user = []
+    item = []
+    score = []
+    with open(filename, 'r') as f:
+        num_samples = 0
+        for line in f:
+            tks = line.strip().split('\t')
+            if len(tks) != 3:
+                continue
+            if (int(tks[0]) > 138000 or int(tks[1]) > 26700):
+                continue
+            num_samples += 1
+            user.append((tks[0]))
+            item.append((tks[1]))
+            score.append((tks[2]))
+    # convert to ndarrays
+    user = mx.nd.array(user, dtype='int32').as_in_context(ctx)
+    item = mx.nd.array(item, dtype='int32').as_in_context(ctx)
+    score = mx.nd.array(score).as_in_context(ctx)
+    # prepare data iters
+    data = {'user': user, 'item': item}
+    label = {'softmax_label': score}
+    iter = mx.io.NDArrayIter(data=data, label=label, batch_size=batch_size)
+    return iter
+
+
+def predict(model, users, items, batch_size=1000, ctx=mx.cpu()):
+    user = mx.nd.array(users, dtype='int32').as_in_context(ctx)
+    item = mx.nd.array(items, dtype='int32').as_in_context(ctx)
+    label = mx.nd.zeros(len(user)).as_in_context(ctx)
+    data = {'user': user, 'item': item}
+    label = {'softmax_label':label}
+    eval_iter = mx.io.NDArrayIter(data=data, label=label, batch_size=batch_size)
+    preds = []
+    for batch in eval_iter:
+        model.forward(batch)
+        mx.nd.waitall()

```

Comment:
waitall is not necessary since asnumpy will call wait_to_read.

Reply:
make sense:)